fem/fem.1
.TH FEM 1 "05 August 2008"
.SH NAME
fem \- Non-Interactive Finite Element Method Solver
.SH SYNOPSIS
\fBfem\fP -i filename [\fIoptions\fP]
.SH DESCRIPTION
\fBfem\fP provides structural analysis (using Finite Element Method).
.SS Options
.TP
\fB-i\fI filename\fR
Name of input file. The input \fBtext file\fP must be in a special form. See \fBfeminput(5)\fP for full description.
.TP
\fB-si\fR
Standard input will be used as input file. If specified, all data files (even specified with \fB-i\fP) will be \fBignored\fP.
.TP
\fB-p\fR
Input data are preparsed (preindexed).  Means that nodes, elements, materials and real sets are pointed by their positions (0..n-1), not by ID numbers. Note that data from preprocessor (\fBtfem\fP, \fBgfem\fP) are always preparsed.
.TP
\fB-o\fI filename\fR
Name of output file (for results).
.TP
\fB-so\fR
Standard output will be used for results. Program messages (errors etc.) will be printed to standard error output.
.TP
\fB-e\fR
Program messages (errors etc.) will be printed to standard error output (not to standard output).
.TP
\fB-r\fI filename\fR
Name of file with results from previous solution.
.TP
\fB-sr\fR
Results from previous solution will be obtained from standard input.
.TP
\fB-ose \fIN\fR
Save every N-th substep (for Newton-Raphson solver only, doesn't work
in MPI mode).
.TP
\fB-osf \fINAME\fR
Template name for substep results file (for Newton-Raphson solver only, doesn't work
in MPI mode). The \fINAME\fR should include full path but \fBnot\fR
the extension and step number (e.g. "/tmp/frame" is correct but
"/tmp/frame002.res" \fBis not\fR).
.TP
\fB-nls\fI N\fR
Selects the non-linear solver (N=0..linear, N=2..Newton-Raphson, N=3..Arc-lenght).
.TP
\fB-nlstp\fI N\fR
Sets number of substeps for the non-linear solver (even number required).
.TP
\fB-nlit\fI N\fR
Sets number of allowed iteration in one substep for the non-linear solver.
.TP
\fB-nlbrk\fR
Solution (Newton-Raphson/Arc-Length) will try continue even after uncoverged step.
.TP
\fB-nlf \fIFILE\fR
Convergence norms will be saved to \fIFILE\fP.
.TP
\fB-d\fR
Run iterative solution for one-directional boundary conditions or materials.
.TP
\fB-t \fINUM\fR
Use parallel solver with \fINUM\fP threads (this option may not be available on your system).
.TP
\fB-tm \fIMIN\fR
Parallelize operations only for vectors and matrices larger than \fIMIN\fP. This option is complementary to the \fB-t \fINUM\fR option and it also may not be available on your system.
.TP
\fB-b\fR 
Exit program after structure stiffness matrix and load vector are assembled.
.TP
\fB-mf \fIFILE\fR 
Save the stiffness matrix of structure to \fIFILE\fP in dense form.
.TP
\fB-msf \fIFILE\fR 
Save the stiffness matrix of structure to \fIFILE\fP in sparse form (compatible with \fImviz(1)\fP program).
.TP
\fB-mmf \fIFILE\fR 
Save the stiffness matrix of structure to \fIFILE\fP in sparse form compatible in a Matrix Market format.
.TP
\fB-vf \fIFILE\fR 
Save the load vector of structure to \fIFILE\fP in dense form.
.TP
\fB-nes\fR 
Uses the Bi-Conjugate Gradient Stabilized Method (with Jacobi
preconditioner) for solution of linear systems.
.TP
\fB-ssor\fR 
Uses the Conjugate Gradient Method with SSOR preconditioner
for solution of linear systems (it also turns off the \fB-nes\fP
functionality).
.TP
\fB-fbc\fR 
Eliminates DOFs with fixed supports (u=0, v=0, w=0). Works only in
statics. Also turnes off computation of reactions in these supports.
Can speed-up solutions with large numbers of suppurts (3D foundation
models).
.TP
\fB-to FILE\fP
Writes temperatures from thermal solution to \fBFILE\fP.
.TP
\fB-tos\fP
Writes temperatures from thermal solution to the standard output.
.TP
\fb-ti FILE\fP
Reads temperatures computed by thermal solution from \fBFILE\fP.
.TP
\fb-tis\fP
Reads temperatures computed by thermal solution from the standard input.
.TP 
\fb-nwm\fP
Runs dynamic analysis by Newmark integration.
.TP
\fB-h\fR
Prints help to screen.
.SH SIGNALS
.TP
\fBSIGINT\fP system signal (can be emmited by \fB<control> <c>\fP key sequence) can be used to terminate correctly (with saving of data files) of Newton-Rahpson or Arc-Lenght Method solution. All other types of solution (including the linear solution) can be only terminated without saving of results.
.SH FILES
.TP
\fBfem\fP
Executable file. Location of file is system-dependent but \fI/usr/local/bin/fem\fP is the most common.
.TP 
\fB$MANPATH/man1/fem.1 \fP
Manual page (this file).
.TP 
\fB$MANPATH/man5/feminput.5 \fP
Manual page for input file.
.SH COMPATIBILITY
Source code has been written in a plain \fBISO C\fP (C89) programming language without any extension.
It should be compiled using any \fIISO C\fP-compliant compiler (at least in theory).
The \fBgcc 3.x\fP compilers usually work.

\fITested\fP systems include:
\fBLinux\fP 2.x (IA32, ARM, PA-RISC),
\fBIRIX\fP 6.5.x (MIPS),
\fBHP-UX\fP 11.x (PA-RISC),
\fBSunOS\fP 5.8, 5.9, 5.10 (UltraSPARC),
\fBAIX\fP 4.3 (IBM POWER4),
\fBMacOS X\fP 10.3 (PowerPC G3/G4/G5),
\fBWindows 2000/XP\fP (IA32, minor issues).
.SH LICENSE
The code is released under the GNU General Public Licence.
.SH AUTHOR
Jiri Brozovsky <jirka@penguin.cz>
.SH "HOME PAGE"
http://fast10.vsb.cz/brozovsky/articles/ufem/index.html
.SH "SEE ALSO"
tgfem(1) mviz(1) feminput(5)
.SH BUGS
.TP
There are tons of bugs. Really!
.TP
Newton-Raphson Solution will work only with even number of steps (2,4,.. 120,..).
.SH VERSION


ChangeLog
Changelog for the user interface only.

See fem/ChangeLog for solver changes.



### Mon Aug 11 17:36:56 CEST 2014

features:
  + menu item for fast reading of single results
  + command rrread for fast readinge of single results

### Mon Aug  4 21:07:09 CEST 2014

features:
  + *extrude may use MAT properties from the original area



### Wed Jul 23 20:43:19 CEST 2014

version 0.2.66

features:
  + CSV output for *list commands
  


### Tue Jul  9 21:53:46 CEST 2013

version 0.2.65

features:
  + triangle elements (11, 18) can be used for automatic regular meshes



### Thu Jan 31 19:11:00 CET 2013

version 0.2.64

features:
  + modal analysis from GUI



### Sun Nov  4 22:07:23 CET 2012

features:
  + user-defined GUI dialogs



### Wed Oct 17 21:05:42 CEST 2012

features:
  + symbols for heats (just a simple dots)


  
### Sat Oct 13 20:34:39 CEST 2012

version 0.2.63

features:
  + updated material 001



### Tue Jun 19 22:37:17 CEST 2012

features:
  + fixed maximal deformation (for animations)
  + fixed color palette (for animations)



### Mon Jun 11 22:13:11 CEST 2012

bugfixes:
  - invalid entities for "gekgen" command



### Sun Jun  3 20:38:20 CEST 2012

features:
  + extrude function for areas and lines
  + enhanced apdl export

bugfixes:
  - iterations number not passed to  solver



### Fri Feb 17 20:48:17 CET 2012

features:
  + coordinate axes symbol in [0,0,0]



### Tue Feb  7 20:44:11 CET 2012

features:
  + aextrude command 

bugfixes:
  - extrude for 4-node areas fixed
  - extrude for 8-node areas fixed



### Sun Feb  5 18:36:35 CET 2012

version 0.2.62

features:
  + new geometry - curvilinear rectangle



### Sun Jan 15 18:59:57 CET 2012

version 0.2.61

features:
  + plotting of deformations (no label, ux,uy,uz only)

bugfixes:
  - Czech documentation
  - missing item label in plns



### Fri Jan  6 21:03:55 CET 2012

version 0.2.58

features:
  + import of tetrahedra mesh from TetGen (tetgen.org)

bugfixes:
  - bad string in "import" command



### Mon Jan  2 19:32:04 CET 2012

version 0.2.58

features:
  + <Enter>, <Esc> can be used in dialogs for easier use



### Sun Nov  6 18:59:39 CET 2011

version 0.2.58

features:
  + two more parameters for material 004 (non-compatible data change)
  + GUI: RS types checked against existing ET in creation dialog



### Mon Oct 17 23:32:20 CEST 2011

features:
  + GUI items for path operations (unstested)



### Thu Aug 11 16:00:26 CEST 2011

version 0.2.57

features:
  + assesment routine for steel links
  + input data for e001, e007 changed (for assesment)



### Mon Jun 27 20:23:05 CEST 2011

version 0.2.56

features:
  + thermal element 019 (tetrahedron)



### Sat Jun 25 20:12:13 CEST 2011

version 0.2.55

features:
  + stationary heat transfer problem (2D only)
  + transfer of temperatures to plane stress model



### Sun Apr 10 20:20:42 CEST 2011

version 0.2.54

features:
  + change of m001 definition



### Thu Jan 13 13:44:18 CET 2011

bugfixes:
  - ininitialized values in data structure (now set to NULL)



### Thu Dec 16 14:33:17 CET 2010

features:
  + multiple paths plotting (untested)



### Mon Nov 29 14:40:35 CET 2010

features:
  + damping parameters  added as random variables



### Mon Nov 29 13:10:31 CET 2010

features:
  + -ssor parameter added to all solver commands



### Tue Nov 16 22:12:33 CET 2010

bugfixes:
  - path filed size problems
  - faults on 64bit Linux fixed (problem related to paths)



### Tue Nov  9 20:05:14 CET 2010

features:
  + fem export adds 0 for dynamics (for compatibility)



### Sun Oct 31 19:26:38 CET 2010

bugfixes:
  - plotting of pathes > 1 was refused



### Sun Oct 31 12:40:11 CET 2010

bugfixes:
  - invalid line width after path plotting



### Sat Oct 23 16:51:27 CEST 2010

features:
  + path picking ("gcreate,pick"_



### Fri Oct 22 19:39:13 CEST 2010

features:
  - zoom for plotted paths

bugfixes:
  - Makefiles (invalid ifeq condition)
  - invalid nodes (and therefore results) in path plotting 



### Thu Oct 21 21:05:28 CEST 2010

features:
  + path nodes can be changed
  + command for selection of active path
  + path plotting command (untested)

bugfixes:
  - "path creation" coredump fixed
  - "pn" command coredump fixed (invalid femInputMaxInt call)
  - "prpath" output fixed



### Wed Oct 20 21:03:49 CEST 2010

version 0.2.53

features:
  + basic commands for path operations (untested)
  + 266 commands



### Mon Oct 18 09:54:18 CEST 2010

features:
  + more path operations functions
	+ cmd_fem splitted to two (cmd_fem, cmd_res)



### Sun Oct 17 16:47:38 CEST 2010

features:
  + added to GitHub repository



### Sat Oct  2 22:19:12 CEST 2010
  + asin (z) and acos (k) parameters in "eval" command
  + femcmd.1 updated



### Mon May  3 09:44:46 CEST 2010

bugfixes:
  - swapped direction of wheel zooming



### Tue Apr 27 21:35:20 CEST 2010

bugfixes:
  - documentation fixes (femcmd.1)



### Sat Oct 31 19:55:10 CET 2009

bugfixes:
  - broken "mirror" commands removed from GUI 



### Mon Aug  3 23:48:56 CEST 2009

features:
  + basic commands for matrix control (untested)



### Thu Jul 30 17:50:30 CEST 2009

features:
  + function for matrix printing (untested)
  + functions for adding and putting values to matrix (untested)



### Tue Jul 28 23:09:26 CEST 2009

features:
  + fdb_math file (matrix field allocations etc)
  + matrix allocation/deallocation commands



### Mon Jul 27 23:30:28 CEST 2009

version 0.2.52

features:
  + "set,first", "set,next" etc is now possible
  + GUI dialog for "set" command
  + GUI "time" dialog shows default time
  + fdb_rops file for result operations (nothink works here)
  + fem_math dependency added (code will follow)



### Sat Jun 27 17:42:16 CEST 2009

bugfixes:
  - element load type initialisation fixed



### Fri Jun 12 23:16:42 CEST 2009

bugfixes:
  - random load numbering fixes (it's only for forces!)



### Thu Jun 11 22:59:02 CEST 2009

features:
  + random variables can include load specified by its ID


fem/fem.c
/*
   File name: fem.c
   Date:      2003/04/07 09:01
   Author:    Jiri Brozovsky

   Copyright (C) 2003 Jiri Brozovsky

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   in a file called COPYING along with this program; if not, write to
   the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
   02139, USA.

	 FEM Solver - main file

  $Id: fem.c,v 1.26 2004/12/31 23:09:25 jirka Exp $
*/

#include "fem.h"
#include "fem_mem.h"
#include "fem_dama.h"
#include "fem_sol.h"
#include "fem_para.h"
#include <string.h>

#ifndef _SMALL_FEM_CODE_
#include "fem_spnt.h"
extern long   femTestConstElem ;
extern long   femTestConstIpn  ;
extern int    femDryRun(void)  ;
extern int    femSolveDynNewmark(double *ofld);
extern int femSolveThermTrans(double *ofld);
#endif
extern long femEigenInvI ;

#ifdef USE_MPI
#include "fem_mpi.h"
#endif


/** Main() function
 */
int main(int argc, char *argv[])
{
	int rv = AF_OK ;

#ifdef RUN_VERBOSE
	msgout = stdout ; /* for output from "fprintf(msgout,...)" */
#endif
	
	if ((rv = fem_parse_params(argc,argv)) != AF_OK) 
	{
#ifdef RUN_VERBOSE
		fprintf(msgout,"[E] %s!\n", _("Command line michmach"));
#endif
		return(rv);
	}

#ifdef USE_MPI /* MPI CODE ---------------------------------- */
	
	/* Setting of data to NULL */
  if ((rv = femDataNull()) != 0) {return(rv);}
  if ((rv = femResNull()) != 0) {return(rv);}

	MPI_Init(&argc, &argv);
	femMPIwhoami(&ppRank, &ppSize);

	if (ppRank == 0)
	{
	  if ((rv = femReadInput(fem_ifile)) != AF_OK) 
		   { exit(rv); }
	}

  if ((rv = femSendInitDataPP()) != AF_OK) 
	{
#ifdef DEVEL_VERBOSE
		fprintf(msgout,"[E] %s %i %s %i\n", _("Data sending error:"),rv,
		_("from process"),ppRank);
#endif
	   exit(rv);
	}

	if (solNoLinS > 1)
	{ 
		if ((rv = femSolveNRM_MPI(2)) != AF_OK) 
		{
#ifdef USE_MPI_LOG
		 printf("Solution returnes: %i\n",rv);
#endif
		 return(rv);
		} 
	}
	else
	{ 
		if ((rv = femSolveMPI()) != AF_OK) 
		{
#ifdef USE_MPI_LOG
		 printf("Solution returnes: %i\n",rv);
#endif
			return(rv);
		} 
	}


#ifndef _SMALL_FEM_CODE_
	femSaPoSave();
	femSaPoDestroy();
#endif

	MPI_Finalize();

#else /* end of MPI CODE ----------------------------------- */
	
	/* Setting of data to NULL */
  if ((rv = femDataNull()) != AF_OK) {return(rv);}
  if ((rv = femResNull()) != AF_OK) {return(rv);}
	if ((rv = femReadInput(fem_ifile)) != AF_OK) {return(rv);}

#ifndef _SMALL_FEM_CODE_
	femSaPoSetNode(0, nLen);
	if ((rv = femSaPoAlloc(solNoLinStepMax, 0)) != AF_OK) {return(rv);}
	femSaPoLen  = solNoLinStepMax ;
#endif


  if (femDynamics == AF_YES) /* dynamics */
  {
    /* modal analysis */
		if (femEigenModal == AF_YES)
		{
      if (femEigenInvI == AF_YES)
      {
    	  rv = femSolveEigenInvIter(1500, 4) ; /* this does not work well */
      }
      else
      {
    	  rv = femSolveEigenLanczos(1500, 4) ; /* this totally fails */
      }
		}
		else
		{
			if (femNewmarkEL == AF_YES) /* Newmark: dynamics */
			{
				rv = femSolveDynNewmark(NULL) ;
			}
			else
			{
#ifdef DEVEL_VERBOSE
				fprintf(msgout,"[E] %s!", _("Unknown analysis type"));
#endif
				return(AF_ERR_VAL);
			}
		}
  }
  else /* statics */
  {
    if ((femTestConstIpn > -1) && (femTestConstElem > -1))
    {
      /* constitutive data testing */
      exit(femDryRun() ) ;
    }
    else /* statics */
    {
      if (femComputePriceOnly == AF_YES) /* computes price of structure */
      {
        return(femComputePrice());
      }
      else
      {
        if (femThermTrans == AF_YES)
        {
	        if ((rv = femSolveThermTrans(NULL)) != AF_OK) {return(rv);} /* thermal transient */
        }
        else
        {
	       if (solNoLinS > 1)
	       {
           switch (solNoLinS)
           {
             case 2: if ((rv = femSolveNRM(0)) != AF_OK) {return(rv);} break;
             case 3: if ((rv = femSolveALM(2)) != AF_OK) {return(rv);} break;
             case 4: if ((rv = femSolveFullALM()) != AF_OK) {return(rv);} break;
             case 6: if ((rv = femSolveDirDisps(0)) != AF_OK) {return(rv);} break;
             default: return(AF_ERR_VAL); break ;
           }
         }
	       else
	         { if ((rv = femSolve()) != AF_OK) {return(rv);} }
        }
      }
    }
  }

#ifndef _SMALL_FEM_CODE_
	femSaPoSave();
	femSaPoDestroy();
#endif

#endif
	
	/* file names */
	if (fem_ifile != NULL); free(fem_ifile); fem_ifile = NULL ;
	if (fem_ofile != NULL); free(fem_ofile); fem_ofile = NULL ;
	if (fem_rfile != NULL); free(fem_rfile); fem_rfile = NULL ;
	if (fem_thrfile != NULL); free(fem_thrfile); fem_thrfile = NULL ;
	if (fem_throfile != NULL); free(fem_throfile); fem_throfile = NULL ;
	return(rv);
}

/* end of fem.c */
